src/modules/hip/kernel/normalize.hpp

```diff
@@ -0,0 +1,1925 @@
+}
+
+__device__ __forceinline__ void reduction_sum_x_hip(float *partialSum_smem)
+{
+    for(uint threadMax = hipBlockDim_x / 2; threadMax >= 1; threadMax /= 2)
+    {
+        if(hipThreadIdx_x < threadMax)
+            partialSum_smem[hipThreadIdx_x] += partialSum_smem[hipThreadIdx_x + threadMax];
+        __syncthreads();
+    }
+}
+
+__global__ void reduce_final_result_hip(float *partialSumTensor,
+                                        uint numPartialSums,
+                                        float *meanTensor,
+                                        float *stdDevTensor,
+                                        bool isMean,
+                                        uint *roiTensor,
+                                        uint axisMask,
+                                        uint numDims)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
+    int id_y = hipBlockIdx_y * hipBlockDim_y + hipThreadIdx_y;
+    int id_z = hipBlockIdx_z * hipBlockDim_z + hipThreadIdx_z;
+    uint *roi = &roiTensor[id_z * numDims * 2 + numDims];
+
+    uint meanFactor;
+    if(numDims == 3)
+    {
+        uint lengthZ = roi[0];
+        uint lengthY = roi[1];
+        uint lengthX = roi[2];
+
+        if(axisMask == 3)
+            meanFactor = lengthZ * lengthY;
+        else if(axisMask == 6)
+            meanFactor = lengthY * lengthX;
+        else if(axisMask == 7)
+            meanFactor = lengthZ * lengthY * lengthX;
+    }
+    else if(numDims == 2)
+    {
+        uint lengthY = roi[0];
+        uint lengthX = roi[1];
+
+        if(axisMask == 2)
+            meanFactor = lengthX;
+        else if(axisMask == 3)
+            meanFactor = lengthY * lengthX;
+    }
+
+    __shared__ float partialSum_smem[16];
+    partialSum_smem[hipThreadIdx_x] = 0.0f;
+
+    float accum = 0.0f;
+    while(id_x < numPartialSums)
+    {
+        uint srcIdx = (id_z * hipGridDim_y * numPartialSums) + (id_y * numPartialSums) + id_x;
+        accum += partialSumTensor[srcIdx];
+        id_x += hipBlockDim_x;
+    }
+    partialSum_smem[hipThreadIdx_x] = accum;
+    __syncthreads();
+
+    // Now do block level reduction sum
+    reduction_sum_x_hip(partialSum_smem);
+
+    // Final store to dst
+    if(hipThreadIdx_x == 0)
+    {
+        if(isMean)
+            meanTensor[id_z * hipGridDim_y + id_y] = partialSum_smem[0] / meanFactor;
+        else
+            stdDevTensor[id_z * hipGridDim_y + id_y] = sqrtf(partialSum_smem[0] / meanFactor);
+    }
+}
+
+__global__ void compute_mean_2d_hip_tensor(float *srcPtr,
+                                           uint2 srcStridesNH,
+                                           float *meanTensor,
+                                           float *partialSumTensor,
+                                           uint *roiTensor,
+                                           uint maxParamVolume,
+                                           uint axisMask)
+{
+    int id_x = hipBlockIdx_x * hipBlockDim_x + hipThreadIdx_x;
```

Comment:
The nd isn't vectorized, but what was the reason for the 2d and 3d specific kernels also not vectorized?

Reply:
Vectorized was taking higher time than the single pixel version. That is the reason the 2D and 3D kernels are not vectorized